src/core/distributions/phylogenetics/tree/birthdeath/GeneralizedLineageHeterogeneousBirthDeathSamplingProcess.cpp

```diff
@@ -0,0 +1,2042 @@
+	for(size_t i = 0; i < obj.size(); ++i)
+	{
+		for(std::map< std::vector<unsigned >, double >::const_iterator it=obj[i].getEventMap().begin(); it != obj[i].getEventMap().end(); ++it)
+		{
+			// Recover the map entries
+			std::vector<unsigned> key = it->first;
+			double val = it->second;
+			// 1-indexing to 0-indexing
+			if ( zero_indexed == false )
+			{
+				for(size_t iK=0; iK<key.size(); ++iK) key[iK] -= 1;
+			}
+			std_object[i][key] = val;
+		}
+	}
+	return std_object;
+}
+
+void GeneralizedLineageHeterogeneousBirthDeathSamplingProcess::updateTree(bool force)
+{
+	if ( force | tree_dirty )
+	{
+		// get the newick string
+		std::string var = this->getValue().getNewickRepresentation();
+
+		if ( use_origin )
+		{
+			// strip out trailing zeros
+			std::string pattern = ":";
+			while ( true )
+			{
+				// if we found a colon stop
+				if ( &var.back() == pattern )
+				{
+					break;
+				}
+				// otherwise, pop off the last character
+				var.pop_back();
+			}
+
+			// now add the tail
+			double origin_age  = age->getValue();
+			double root_age    = this->getValue().getRoot().getAge();
+			double tail_length = origin_age - root_age;
+			var += std::to_string(tail_length);
+		}
+		else
+		{
+			// strip off the tail
+			// strip out trailing zeros
+			std::string pattern = ":";
+			while ( true )
+			{
+				// if we found a colon stop
+				if ( &var.back() == pattern )
+				{
+					break;
+				}
+				// otherwise, pop off the last character
+				var.pop_back();
+			}
+
+			// now add the trailing zeros
+			var += "0.00000";
+		}
+
+		// make sure there's a closing semicolon
+		var += ";";
+
+		// set the tree
+		tp_ptr->setTree(var);
+
+	}
+}
+
+void GeneralizedLineageHeterogeneousBirthDeathSamplingProcess::updateData(bool force)
+{
+	if ( force )
+	{
+		// get the data
+		const RevBayesCore::AbstractHomologousDiscreteCharacterData &dat = getCharacterData();
+
+		size_t num_taxa = taxa.size();
+		std::map< std::string, std::vector<double> > taxon_map;
+		std::vector<std::string> taxon_names(num_taxa);
+		for(size_t i = 0; i < num_taxa; ++i)
+		{
+			const AbstractDiscreteTaxonData& this_taxon_data = dat.getTaxonData(taxa[i].getName());
+
+			// if the data are weighted, use the weights
+			if ( this_taxon_data[0].isWeighted() )
+			{
+				taxon_map[taxa[i].getName()] = this_taxon_data[0].getWeights();
+			}
+			else
+			{
+				RbBitSet this_bit_set = this_taxon_data[0].getState();
+				std::vector<double> this_taxon_states(num_states);
+				for(size_t j = 0; j < num_states; ++j)
+				{
+					this_taxon_states[j] = this_bit_set[j] == true ? 1.0 : 0.0;
+				}
+				taxon_map[taxa[i].getName()] = this_taxon_states;
+			}
+
+			taxon_names[i] = taxa[i].getName();
+		}
+
+		// set the data
+		tp_ptr->setData(taxon_names, taxon_map);
+
+	}
+}
+
+void GeneralizedLineageHeterogeneousBirthDeathSamplingProcess::updateRootFrequency(bool force)
+{
+	if ( force | root_frequency_dirty )
```

Comment:
This is a bitwise or (one `|`) but should be a logical or (`||`).
This is not just a style issue -- sometimes this leads to weird results because `|` has different precedence, and also causes compiler warnings.